Use cp rather than ln -s for node client

matklad opened the pull request:
Our NPM package needs to include tigerbeetle source code (.zig files), because we are building the actual node plugin on the user's machine.

Today, we include that code by symlinkin (creating a cyclical structure). This used to work with npm 8, but broke with npm 9.

See <https://github.com/tigerbeetledb/tigerbeetle/pull/437>

Judging by <https://github.com/npm/npm/issues/3310>, npm semi-intentionally does not want to support symlinks (which is a position I can understand).

So let's do something marginally less hacky, by copying the files over rather than linking them. We can do that in `prepare` hook, which is executed on local `nmp i`, `npm pack` and `npm publish`.

## Pre-merge checklist

Performance:

* [ ] Compare `zig benchmark` on linux before and after this pr.
    ``` sh
    # benchmark results before
    ...
    
    # benchmark results after
    ...
    ```
OR
* [x] I am very sure this PR could not affect performance.

matklad commented on src/clients/node/scripts/build.sh:
```diff
@@ -2,7 +2,7 @@
 
  set -e
 
-docker run -v "$(pwd)/../../..":/wrk -w /wrk/src/clients/node --entrypoint bash node:18 -c "
```
I wonder what we should pick here?

- on the one hand, we are shipping a library, not an app, so it's advantegous to make this a rolling versions to catch cases where npm regresses us
- on the other hand, if we are doign that, we probably want latest node, rather than just lts
- on the third hand, this stuff is for _building_ the package for publishing, so arguably only we should worry about this, and not our users, and it might be better to pin?

eatonphil replied:
Eventually we probably need a matrix of versions. So I don't think it really matters what we pick at the moment.

eatonphil commented on src/clients/node/package.json:
```diff
@@ -36,7 +34,7 @@
     "benchmark": "./scripts/benchmark.sh",
     "test": "node dist/test",
-    "postinstall": "npm run install_zig && npm run download_node_headers && npm run build_lib",
+    "postinstall": "test -d src/tigerbeetle || (mkdir -p src/tigerbeetle && cp -r ../../../{scripts,src/{*.zig,io,lsm,testing,vsr,state_machine}} src/tigerbeetle) && npm run install_zig && npm run download_node_headers && npm run build_lib",
```
If you put this in a script we can shellcheck it. Shellcheck will also validate that all the commands and syntax you used work in posix sh (alpine doesn't have bash, nor do we require it).

eatonphil replied:
But you'd need to add that file to the `files` list in package.json.

eatonphil replied:
The alpine build failure is likely because of bashisms

matklad replied:
Yeah, this in generall going to be a bite more complicated/ugly than I thought. Turns out I can't use `prepare` hook for this, cause it runs after install. 

So I think what we could do is indeed put the logic into a script, run it on postinstall, but make the script _itself_ conditional such that it only does the copy if in the repo...

sentientwaffle replied:
Yeah, busybox/posix shell doesn't support `{a,b,c}`.

matklad committed:
fix compat with new npm

Previously, we symlinked from within clients/node to the root of the
repo to make it possible for npm to include `.zig` files into npm
package (so that we can compile these zig files on the user's machine).

npm 9 stopped including symlinks, so here we switch to `cp` instead.

We do this in a slightly hacky way, by running `cp` from the script
which builds the node plugin. When this script is run inside the repo
(eg, before publishing), it copies the files over. When the script is
run on the user's machine, it does nothing.

This setup helps working on the code, as there's no explicit "copy" step
you can forget to run.

matklad commented:
@eatonphil any idea what gives me access denied here?

https://github.com/tigerbeetledb/tigerbeetle/actions/runs/4105490381/jobs/7082420770

Works fine locally...

matklad commented:
Ok, debugged this: 

```
mkdirat(AT_FDCWD, "/root/.cache/zig", 0755) = -1 EACCES (Permission denied)
```

So, zig fails to create a global cache

matklad commented on src/clients/node/scripts/build_lib.sh:
```diff
@@ -24,11 +43,10 @@ else
 # Need to do string eval-ing because of shellcheck's strict string
 # interpolation rules.
 cmd="./zig/zig build-lib \
+	--global-cache-dir ./zig/global-cache \
```
After I moved `install_zig` from a separate npm script to be part of `build_lib`, I got access denied error on cache. I don't understand what was the cause of that difference, but setting global cache to a local dir here helps. 

It also seems like a generally right thing to do -- we don't want to pollute any global dirs with our stuff, zig should be an impl detail of the node package.

matklad commented:
After a long struggle between humanity on the one side and node&docker on the other, humanity won, so this is ready for review!

sentientwaffle commented:
bors r+
